src/Storages/Statistic/Estimator.cpp

```diff
@@ -0,0 +1,135 @@
+#include <Storages/Statistic/Estimator.h>
+#include <Storages/MergeTree/RPNBuilder.h>
+
+namespace DB
+{
+
+std::optional<std::string> ConditionEstimator::extractSingleColumn(const RPNBuilderTreeNode & node) const
+{
+    if (node.isConstant())
+    {
+        return std::nullopt;
+    }
+
+    if (!node.isFunction())
+    {
+        auto column_name = node.getColumnName();
+        return {column_name};
+    }
+
+    auto function_node = node.toFunctionNode();
+    size_t arguments_size = function_node.getArgumentsSize();
+    std::optional<std::string> result;
+    for (size_t i = 0; i < arguments_size; ++i)
+    {
+        auto function_argument = function_node.getArgumentAt(i);
+        auto subresult = extractSingleColumn(function_argument);
+        if (subresult == std::nullopt)
+            continue;
+        else if (subresult == "")
+            return "";
+        else if (result == std::nullopt)
+            result = subresult;
+        else if (result.value() != subresult.value())
+            return "";
```

Comment:
It looks a bit messy. Can we leave only returning of `std::nullopt` in if there is no suitable column?